Initial description of annotation collection

handrews opened the pull request:
***NOTE:*** _This PR uses language added in PR #609, which was recently merged._

This covers the basics of annotation collection (#530), and its interaction
with assertions and applicators.  It does not get into the exact
output structure for annotation results, which is intentional.
The goal is to establish the process, and then nail down the
format once the process is solid.

awwright commented on jsonschema-core.xml:
```diff
@@ -1066,6 +1066,94 @@
+            <t>
+                Annotations are collected by keywords that explicitly define
+                annotation-collecting behavior.  Note that boolean schemas cannot
+                produce annotations as they do not make use of keywords.
```
Is this not obvious enough?

awwright commented on jsonschema-core.xml:
```diff
@@ -1066,6 +1066,94 @@
+                A collected annotation MUST include the following information:
+                <list>
+                    <t>
+                        The name of the annotation, which MUST be identical to the keyword
```
How about simply "The name of the keyword that produced the annotation"

handrews replied:
Never mind, I see what you mean.  Will update.

awwright commented on jsonschema-core.xml:
```diff
@@ -1066,6 +1066,94 @@
+                </list>
+            </t>
+            <t>
+                If multiple schema locations attach values to the same instance location,
```
What's an "instance location"? Maybe just "instance"?

awwright commented on jsonschema-core.xml:
```diff
@@ -1066,6 +1066,94 @@
+                For example, one application may consider a "description" annotation that is
+                in the same schema object as a "$ref" to override any "description" in the
+                reference's target schema.  A different application may prefer to concatenate
+                all "description" annotations based on whatever ordering it defines.
```
If we're going to specify sometimes annotations can override one another, there should be a deterministic algorithm defined. Consider how CSS ranks which property is highest priority.

Presenting all the generated values as equals is probably the most theoretically pure way, but I'm not sure how practical this is for schema authors. I don't think I'll have a solid opinion on this until we see implementations.

handrews replied:
@awwright I'm going to add an example to make it more clear that we're not just punting on the question of precedence, but rather providing tools for applications to define their own in a clear way.

awwright commented on jsonschema-core.xml:
```diff
@@ -1066,6 +1066,94 @@
+                    If any keyword in a schema object produces a false assertion
+                    result, then all annotations from all keywords in that schema
+                    object, and any of its subschemas or referenced schemas, MUST
+                    be discarded.
```
This could be shortened for clarity, I think.
"Annotations are only generated from instances that validate against the schema."

And care is needed so we don't confuse people implementing "not". The note about "not" is only (and should only be) non-normative.

handrews replied:
@awwright I'm going to remove most of the "Annotations and Applicators" section to get rid of the details around `not` and similar cases, and try to clarify why I'm using the "must be discarded" language in the "Annotations and Assertions" section.

handrews commented:
@awwright I've made several updates and tried to incorporate most of your feedback.  I've removed a number of potentially confusing details- hopefully people will understand how they follow from the simpler descriptions.

I also added an example of how an application might make use of multiple values.  Which hopefully will make the point and usefulness more clear.  If you still really want the spec to define behavior for every keyword, let's split that out into a new issue (in which case I will remove the controversial language from this PR so that we can merge the uncontroversial parts).

The two things I did not change are:

* the line about boolean schemas.  I agree that it should be obvious, but since boolean schemas produce assertion results despite not having keywords, I feel like it needs to be explicit that they do not produce annotation results.  Even though I can't think of what they would possibly produce.

* "instance location", which is terminology that, along with "schema location", I've been using extensively across multiple PRs.  It appears in all three documents at this point.  If you think there is a serious problem with it, please file a new issue and we will work it out.  But it is not specific to this PR.

handrews commented on jsonschema-core.xml:
```diff
@@ -1140,6 +1140,191 @@
+                A collected annotation MUST include the following information:
+                <list>
+                    <t>
+                        The name of the keyword that defines the annotation
```
Yeah, I'll change that and then commit, thanks!

vearutop commented:
If [validation error response](https://github.com/json-schema-org/json-schema-spec/issues/396) is  also powered by annotations, then boolean schema that produces the error should be in annotation, shouldn't it?
